### PR TITLE
[SPARK-37160][SQL] Add a config to optionally disable padding for char type

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Adds a new config to optionally disable padding for char type, by replacing char type with varchar type in the table schema when creating/updating tables. Existing tables with char type column/field are not affected.

### Why are the changes needed?

char type is a legacy in the SQL world. Its padding behavior was mostly for storage optimization at that time, which is not a useful SQL semantic. As result,  [MySQL trims the spaces for char type by default](https://docs.oracle.com/cd/E17952_01/mysql-8.0-en/sql-mode.html#sqlmode_pad_char_to_full_length), [Snowflake simply treats char as an alias of varchar](https://docs.snowflake.com/en/sql-reference/data-types-text.html#char-character).

It makes sense for Spark to also demote char type by adding this config.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala [hunk 1]
```diff
@@ -138,6 +154,21 @@ object CharVarcharUtils extends Logging {
     StructType(fields)
   }
 
+  def getRawSchema(schema: StructType, conf: SQLConf): StructType = {
+    val fields = schema.map { field =>
+      getRawType(field.metadata).map { dt =>
+        if (conf.getConf(SQLConf.CHAR_AS_VARCHAR)) {
+          val metadata = new MetadataBuilder().withMetadata(field.metadata)
+            .remove(CHAR_VARCHAR_TYPE_STRING_METADATA_KEY).build()
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:162]: Isn't the metadata used for char and varchar to string conversion? If `CHAR_AS_VARCHAR` is enabled, it means that we also need drop the metadata even the raw type is varchar?

**Review by viirya — COMMENTED** [review]
Looks good. Just one question regarding the interaction with char and varchar to string conversion.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:162]: Here we are getting the raw schema, which is restored from this special metadata, and we don't need this special metadata anymore.

It's usually no harm to still keep this special metadata, but in this case, it's inconsistent with the returned raw schema, and it's safer to remove it. I can do the same in the old `getRawSchema` as well if you think it's clearer.

**Review by yaooqinn — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:162]: Okay, then I think it is fine.

**Review by viirya — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Pull request closed without merging**